File: pkg/java/src/main/java/dev/openfga/language/JsonToDslTransformer.java
package dev.openfga.language;

import static java.util.Objects.requireNonNullElseGet;
import static java.util.stream.Collectors.joining;

import com.fasterxml.jackson.core.JsonProcessingException;
import dev.openfga.language.errors.UnsupportedDSLNestingException;
import dev.openfga.language.errors.UnsupportedModularModulesException;
import dev.openfga.sdk.api.model.*;
import java.util.*;
import java.util.function.Function;
import java.util.stream.IntStream;

public class JsonToDslTransformer {

    private static final String EOL = System.lineSeparator();

    private boolean includeSourceInformation = false;

    public String transform(String json) throws JsonProcessingException {
        return transform(json, false);
    }

    public String transform(String json, boolean includeSourceInformation) throws JsonProcessingException {
        this.includeSourceInformation = includeSourceInformation;
        var model = JSON.parse(json, AuthorizationModel.class);
        return transformJSONToDSL(model);
    }

    private String transformJSONToDSL(AuthorizationModel model) {
        var schemaVersion = "1.1";
        if (model != null && model.getSchemaVersion() != null) {
            schemaVersion = model.getSchemaVersion();
        }

        var modular = isModularModel(model);

        var formattedTypeDefinitions = new StringBuilder();
        if (model != null && model.getTypeDefinitions() != null) {
            var typeDefinitions = new ArrayList<>(model.getTypeDefinitions());
            if (modular) {
                typeDefinitions.sort((a, b) -> sortByModule(
                        a.getType(),
                        b.getType(),
                        module(a.getMetadata()),
                        module(b.getMetadata()),
                        file(a.getMetadata()),
                        file(b.getMetadata())));
            }
            for (var typeDefinition : typeDefinitions) {
                formattedTypeDefinitions
                        .append(formatType(typeDefinition, modular))
                        .append(EOL);
            }
        }

        var fomattedConditions = formatConditions(model, modular);

        return "model" + EOL + "  schema " + schemaVersion + EOL + formattedTypeDefinitions + fomattedConditions;
    }

    private String formatType(TypeDefinition typeDef, boolean modular) {
        var typeName = typeDef.getType();
        var metadata = typeDef.getMetadata();
        var formatedTypeBuilder = new StringBuilder(EOL)
                .append("type ")
                .append(typeName)
                .append(constructSourceComment(module(metadata), file(metadata), ""));

        var relations = requireNonNullElseGet(typeDef.getRelations(), HashMap<String, Userset>::new);
        var emptyMetadataRelation = new HashMap<String, RelationMetadata>();
        var metadataRelations = metadata != null ? metadata.getRelations() : emptyMetadataRelation;
        if (metadataRelations == null) {
            metadataRelations = emptyMetadataRelation;
        }

        if (!relations.isEmpty()) {
            formatedTypeBuilder.append(EOL).append("  relations");
            var relationNames = new ArrayList<>(relations.keySet());
            if (modular) {
                var meta = metadataRelations;
                relationNames.sort((a, b) -> sortByModule(
                        a, b, module(meta.get(a)), module(meta.get(b)), file(meta.get(a)), file(meta.get(b))));
            }
            for (var relationName : relationNames) {
                var formattedRelationString = formatRelation(
                        typeName, relationName, relations.get(relationName), metadataRelations.get(relationName));
                formatedTypeBuilder.append(EOL).append(formattedRelationString);
            }
        }

        return formatedTypeBuilder.toString();
    }

    private String formatRelation(
            String typeName, String relationName, Userset relationDefinition, RelationMetadata relationMetadata) {
        var validator = new DirectAssignmentValidator();

        var typeRestrictions = requireNonNullElseGet(
                relationMetadata.getDirectlyRelatedUserTypes(), ArrayList<RelationReference>::new);

        RelationFormatter formatter = this::formatSubRelation;

        if (relationDefinition.getDifference() != null) {
            formatter = this::formatDifference;
        } else if (relationDefinition.getUnion() != null) {
            formatter = this::formatUnion;
        } else if (relationDefinition.getIntersection() != null) {
            formatter = this::formatIntersection;
        }

        var formattedRelation =
                formatter.format(typeName, relationName, relationDefinition, typeRestrictions, validator);
        if (validator.occurences() == 0
                || (validator.occurences() == 1 && validator.isFirstPosition(relationDefinition))) {
            var sourceComment =
                    constructSourceComment(module(relationMetadata), file(relationMetadata), " extended by:");
            return "    define " + relationName + ": " + formattedRelation + sourceComment;
        }

        throw new UnsupportedDSLNestingException(typeName, relationName);
    }

    private StringBuilder formatDifference(
            String typeName,
            String relationName,
            Userset relationDefinition,
            List<RelationReference> typeRestrictions,
            DirectAssignmentValidator validator) {
        var base = formatSubRelation(
                typeName, relationName, relationDefinition.getDifference().getBase(), typeRestrictions, validator);
        var difference = formatSubRelation(
                typeName, relationName, relationDefinition.getDifference().getSubtract(), typeRestrictions, validator);
        return new StringBuilder(base).append(" but not ").append(difference);
    }

    private StringBuilder formatUnion(
            String typeName,
            String relationName,
            Userset relationDefinition,
            List<RelationReference> typeRestrictions,
            DirectAssignmentValidator validator) {
        return joinChildren(
                Userset::getUnion, "or", typeName, relationName, relationDefinition, typeRestrictions, validator);
    }

    private StringBuilder formatIntersection(
            String typeName,
            String relationName,
            Userset relationDefinition,
            List<RelationReference> typeRestrictions,
            DirectAssignmentValidator validator) {
        return joinChildren(
                Userset::getIntersection,
                "and",
                typeName,
                relationName,
                relationDefinition,
                typeRestrictions,
                validator);
    }

    private StringBuilder joinChildren(
            Function<Userset, Usersets> childrenAccessor,
            String operator,
            String typeName,
            String relationName,
            Userset relationDefinition,
            List<RelationReference> typeRestrictions,
            DirectAssignmentValidator validator) {
        List<Userset> children = null;
        if (relationDefinition != null && childrenAccessor.apply(relationDefinition) != null) {
            children = prioritizeDirectAssignment(
                    childrenAccessor.apply(relationDefinition).getChild());
        }
        children = requireNonNullElseGet(children, ArrayList::new);

        var formattedUnion = new StringBuilder();
        boolean notFirst = false;
        for (var child : children) {
            if (notFirst) {
                formattedUnion.append(" ").append(operator).append(" ");
            } else {
                notFirst = true;
            }
            formattedUnion.append(formatSubRelation(typeName, relationName, child, typeRestrictions, validator));
        }

        return formattedUnion;
    }

    private static List<Userset> prioritizeDirectAssignment(List<Userset> usersets) {
        if (usersets != null && !usersets.isEmpty()) {
            var thisPosition = IntStream.range(0, usersets.size())
                    .filter(i -> usersets.get(i).getThis() != null)
                    .findFirst()
                    .orElse(-1);
            if (thisPosition > 0) {
                var thisUserset = usersets.remove(thisPosition);
                usersets.add(0, thisUserset);
            }
        }

        return usersets;
    }
    ;

    private static class DirectAssignmentValidator {
        private int occured = 0;

        void incr() {
            occured++;
        }

        int occurences() {
            return occured;
        }

        public boolean isFirstPosition(Userset userset) {
            if (userset.getThis() != null) {
                return true;
            }

            if (userset.getDifference() != null && userset.getDifference().getBase() != null) {
                if (userset.getDifference().getBase().getThis() != null) {
                    return true;
                } else {
                    return isFirstPosition(userset.getDifference().getBase());
                }
            } else if (userset.getIntersection() != null
                    && userset.getIntersection().getChild() != null
                    && !userset.getIntersection().getChild().isEmpty()) {
                if (userset.getIntersection().getChild().get(0).getThis() != null) {
                    return true;
                } else {
                    return isFirstPosition(userset.getIntersection().getChild().get(0));
                }
            } else if (userset.getUnion() != null
                    && !userset.getUnion().getChild().isEmpty()) {
                if (userset.getUnion().getChild().get(0).getThis() != null) {
                    return true;
                } else {
                    return isFirstPosition(userset.getUnion().getChild().get(0));
                }
            }
            return false;
        }
    }

    private CharSequence formatSubRelation(
            String typeName,
            String relationName,
            Userset relationDefinition,
            List<RelationReference> typeRestrictions,
            DirectAssignmentValidator validator) {
        if (relationDefinition.getThis() != null) {
            validator.incr();
            return formatThis(typeRestrictions);
        }

        if (relationDefinition.getComputedUserset() != null) {
            return formatComputedUserset(relationDefinition);
        }

        if (relationDefinition.getTupleToUserset() != null) {
            return formatTupleToUserset(relationDefinition);
        }

        if (relationDefinition.getUnion() != null) {
            return formatUnion(typeName, relationName, relationDefinition, typeRestrictions, validator)
                    .insert(0, '(')
                    .append(')');
        }

        if (relationDefinition.getIntersection() != null) {
            return formatIntersection(typeName, relationName, relationDefinition, typeRestrictions, validator)
                    .insert(0, '(')
                    .append(')');
        }

        if (relationDefinition.getDifference() != null) {
            return formatDifference(typeName, relationName, relationDefinition, typeRestrictions, validator)
                    .insert(0, '(')
                    .append(')');
        }

        throw new UnsupportedDSLNestingException(typeName, relationName);
    }

    private CharSequence formatThis(List<RelationReference> typeRestrictions) {
        return requireNonNullElseGet(typeRestrictions, ArrayList<RelationReference>::new).stream()
                .map(this::formatTypeRestriction)
                .collect(joining(", ", "[", "]"));
    }

    private CharSequence formatTypeRestriction(RelationReference restriction) {
        var typeName = restriction.getType();
        var relation = restriction.getRelation();
        var wildcard = restriction.getWildcard();
        var condition = restriction.getCondition();

        var formattedTypeRestriction = new StringBuilder(typeName);

        if (wildcard != null) {
            formattedTypeRestriction.append(":*");
        }

        if (relation != null && !relation.isEmpty()) {
            formattedTypeRestriction.append('#').append(relation);
        }

        if (condition != null && !condition.isEmpty()) {
            formattedTypeRestriction.append(" with ").append(condition);
        }
        return formattedTypeRestriction;
    }

    private CharSequence formatComputedUserset(Userset relationDefinition) {
        return relationDefinition.getComputedUserset().getRelation();
    }

    private CharSequence formatTupleToUserset(Userset relationDefinition) {
        String computedUserset = "";
        String tupleset = "";
        if (relationDefinition != null && relationDefinition.getTupleToUserset() != null) {
            if (relationDefinition.getTupleToUserset().getComputedUserset() != null) {
                computedUserset = relationDefinition
                        .getTupleToUserset()
                        .getComputedUserset()
                        .getRelation();
            }
            if (relationDefinition.getTupleToUserset().getTupleset() != null) {
                tupleset = relationDefinition.getTupleToUserset().getTupleset().getRelation();
            }
        }
        return new StringBuilder(computedUserset).append(" from ").append(tupleset);
    }

    private CharSequence formatConditions(AuthorizationModel model, boolean modular) {
        var conditions = model.getConditions();
        if (conditions == null || conditions.isEmpty()) {
            return "";
        }

        var conditionNames = new ArrayList<>(conditions.keySet());
        if (modular) {
            conditionNames.sort((a, b) -> sortByModule(
                    a,
                    b,
                    module(conditions.get(a).getMetadata()),
                    module(conditions.get(b).getMetadata()),
                    file(conditions.get(a).getMetadata()),
                    file(conditions.get(b).getMetadata())));
        } else {
            conditionNames.sort(null);
        }

        var formattedConditions = new StringBuilder();
        for (var conditionName : conditionNames) {
            formattedConditions.append(EOL).append(formatCondition(conditionName, conditions.get(conditionName)));
        }

        return formattedConditions;
    }

    private CharSequence formatCondition(String conditionName, Condition conditionDef) {
        if (!conditionName.equals(conditionDef.getName())) {
            throw new IllegalArgumentException("conditionName must match condition.getName()");
        }

        var formattedParameters = formatConditionParameters(conditionDef.getParameters());
        return new StringBuilder("condition ")
                .append(conditionDef.getName())
                .append('(')
                .append(formattedParameters)
                .append(") {")
                .append(EOL)
                .append("  ")
                .append(conditionDef.getExpression())
                .append(EOL)
                .append('}')
                .append(constructSourceComment(
                        module(conditionDef.getMetadata()), file(conditionDef.getMetadata()), ""))
                .append(EOL);
    }

    private CharSequence formatConditionParameters(Map<String, ConditionParamTypeRef> parameters) {
        if (parameters == null || parameters.isEmpty()) {
            return "";
        }

        return new TreeMap<>(parameters)
                .entrySet().stream()
                        .map(entry -> {
                            var parameterName = entry.getKey();
                            var parameterType = entry.getValue();
                            var formattedParameterType = parameterType
                                    .getTypeName()
                                    .getValue()
                                    .replace("TYPE_NAME_", "")
                                    .toLowerCase();
                            if (formattedParameterType.equals("list") || formattedParameterType.equals("map")) {
                                var genericTypeString = parameterType
                                        .getGenericTypes()
                                        .get(0)
                                        .getTypeName()
                                        .getValue()
                                        .replace("TYPE_NAME_", "")
                                        .toLowerCase();
                                formattedParameterType = formattedParameterType + "<" + genericTypeString + ">";
                            }
                            return new StringBuilder(parameterName).append(": ").append(formattedParameterType);
                        })
                        .collect(joining(", "));
    }

    private interface RelationFormatter {
        CharSequence format(
                String typeName,
                String relationName,
                Userset relationDefinition,
                List<RelationReference> typeRestrictions,
                DirectAssignmentValidator validator);
    }

    private CharSequence constructSourceComment(String module, String file, String leadingString) {
        if (!includeSourceInformation || (isBlank(module) && isBlank(file))) {
            return "";
        }
        return " #" + leadingString + " module: " + nullToEmpty(module) + ", file: " + nullToEmpty(file);
    }

    private static boolean isModularModel(AuthorizationModel model) {
        if (model == null || model.getTypeDefinitions() == null) {
            return false;
        }
        return model.getTypeDefinitions().stream().anyMatch(typeDef -> !isBlank(module(typeDef.getMetadata())));
    }

    /** Returns the alphabetically sorted set of modules declared in a modular model, or empty if not modular. */
    public static List<String> getModulesFromJSON(AuthorizationModel model) {
        var schemaVersion = model != null && model.getSchemaVersion() != null ? model.getSchemaVersion() : "1.1";
        if (!schemaVersion.equals("1.2")) {
            throw new UnsupportedModularModulesException(schemaVersion);
        }

        if (!isModularModel(model)) {
            return List.of();
        }

        var modules = new TreeSet<String>();
        for (var typeDef : model.getTypeDefinitions()) {
            var metadata = typeDef.getMetadata();
            if (metadata == null) {
                continue;
            }
            if (!isBlank(metadata.getModule())) {
                modules.add(metadata.getModule());
            }
            if (metadata.getRelations() != null) {
                metadata.getRelations().values().stream()
                        .map(RelationMetadata::getModule)
                        .filter(m -> !isBlank(m))
                        .forEach(modules::add);
            }
        }

        if (model.getConditions() != null) {
            model.getConditions().values().stream()
                    .map(condition -> module(condition.getMetadata()))
                    .filter(m -> !isBlank(m))
                    .forEach(modules::add);
        }

        return new ArrayList<>(modules);
    }

    private static int sortByModule(
            String aName, String bName, String aModule, String bModule, String aFile, String bFile) {
        if (isBlank(aModule) && isBlank(bModule)) {
            return aName.compareTo(bName);
        }
        if (isBlank(aModule)) {
            return -1;
        }
        if (isBlank(bModule)) {
            return 1;
        }
        if (!aModule.equals(bModule)) {
            return aModule.compareTo(bModule);
        }
        if (!nullToEmpty(aFile).equals(nullToEmpty(bFile))) {
            return nullToEmpty(aFile).compareTo(nullToEmpty(bFile));
        }
        return aName.compareTo(bName);
    }

    private static String module(Metadata metadata) {
        return metadata == null ? null : metadata.getModule();
    }

    private static String module(RelationMetadata metadata) {
        return metadata == null ? null : metadata.getModule();
    }

    private static String module(ConditionMetadata metadata) {
        return metadata == null ? null : metadata.getModule();
    }

    private static String file(Metadata metadata) {
        return metadata == null || metadata.getSourceInfo() == null
                ? null
                : metadata.getSourceInfo().getFile();
    }

    private static String file(RelationMetadata metadata) {
        return metadata == null || metadata.getSourceInfo() == null
                ? null
                : metadata.getSourceInfo().getFile();
    }

    private static String file(ConditionMetadata metadata) {
        return metadata == null || metadata.getSourceInfo() == null
                ? null
                : metadata.getSourceInfo().getFile();
    }

    private static boolean isBlank(String value) {
        return value == null || value.isEmpty();
    }

    private static String nullToEmpty(String value) {
        return value == null ? "" : value;
    }
}

File: pkg/csharp/Transformers/JsonToDslTransformer.cs
using System.Text;
using OpenFga.Sdk.Model;
using OpenFgaLanguage.Errors;
using OpenFgaLanguage.Utils;

namespace OpenFgaLanguage.Transformers;

public class JsonToDslTransformer
{
    private static readonly string EOL = Environment.NewLine;

    public string Transform(string json)
    {
        AuthorizationModel? model = null;
        if (!string.IsNullOrEmpty(json) && json != "null")
        {
            model = Json.Parse<AuthorizationModel>(json);
        }
        return TransformJsonToDsl(model);
    }

    private string TransformJsonToDsl(AuthorizationModel? model)
    {
        var schemaVersion = "1.1";
        if (model?.SchemaVersion != null)
        {
            schemaVersion = model.SchemaVersion;
        }

        var formattedTypeDefinitions = new StringBuilder();
        if (model?.TypeDefinitions != null)
        {
            foreach (var typeDefinition in model.TypeDefinitions)
            {
                formattedTypeDefinitions.Append(FormatType(typeDefinition)).Append(EOL);
            }
        }

        var formattedConditions = FormatConditions(model);

        return $"model{EOL}  schema {schemaVersion}{EOL}{formattedTypeDefinitions}{formattedConditions}";
    }

    private string FormatType(TypeDefinition typeDef)
    {
        var typeName = typeDef.Type;
        var formattedTypeBuilder = new StringBuilder(EOL).Append("type ").Append(typeName);

        var relations = typeDef.Relations ?? new Dictionary<string, Userset>();
        var metadata = typeDef.Metadata;
        var emptyMetadataRelation = new Dictionary<string, RelationMetadata>();
        var metadataRelations = metadata?.Relations ?? emptyMetadataRelation;

        if (relations.Count > 0)
        {
            formattedTypeBuilder.Append(EOL).Append("  relations");
            foreach (var relationEntry in relations)
            {
                var relationName = relationEntry.Key;
                var relationDefinition = relationEntry.Value;
                var formattedRelationString = FormatRelation(typeName, relationName, relationDefinition, 
                    metadataRelations.GetValueOrDefault(relationName));
                formattedTypeBuilder.Append(EOL).Append(formattedRelationString);
            }
        }

        return formattedTypeBuilder.ToString();
    }

    private string FormatRelation(string typeName, string relationName, Userset relationDefinition, RelationMetadata? relationMetadata)
    {
        var validator = new DirectAssignmentValidator();

        var typeRestrictions = relationMetadata?.DirectlyRelatedUserTypes ?? new List<RelationReference>();

        Func<string, string, Userset, List<RelationReference>, DirectAssignmentValidator, StringBuilder> formatter = FormatSubRelation;

        if (relationDefinition.Difference != null)
        {
            formatter = FormatDifference;
        }
        else if (relationDefinition.Union != null)
        {
            formatter = FormatUnion;
        }
        else if (relationDefinition.Intersection != null)
        {
            formatter = FormatIntersection;
        }

        var formattedRelation = formatter(typeName, relationName, relationDefinition, typeRestrictions, validator);
        
        if (validator.Occurrences() == 0 || (validator.Occurrences() == 1 && validator.IsFirstPosition(relationDefinition)))
        {
            return $"    define {relationName}: {formattedRelation}";
        }

        throw new UnsupportedDSLNestingException(typeName, relationName);
    }

    private StringBuilder FormatDifference(string typeName, string relationName, Userset relationDefinition, 
        List<RelationReference> typeRestrictions, DirectAssignmentValidator validator)
    {
        var baseRelation = FormatSubRelation(typeName, relationName, relationDefinition.Difference.Base, typeRestrictions, validator);
        var difference = FormatSubRelation(typeName, relationName, relationDefinition.Difference.Subtract, typeRestrictions, validator);
        return new StringBuilder(baseRelation.ToString()).Append(" but not ").Append(difference);
    }

    private StringBuilder FormatUnion(string typeName, string relationName, Userset relationDefinition, 
        List<RelationReference> typeRestrictions, DirectAssignmentValidator validator)
    {
        return JoinChildren(relationDefinition.Union?.Child, "or", typeName, relationName, relationDefinition, typeRestrictions, validator);
    }

    private StringBuilder FormatIntersection(string typeName, string relationName, Userset relationDefinition, 
        List<RelationReference> typeRestrictions, DirectAssignmentValidator validator)
    {
        return JoinChildren(relationDefinition.Intersection?.Child, "and", typeName, relationName, relationDefinition, typeRestrictions, validator);
    }

    private StringBuilder JoinChildren(List<Userset>? children, string op, string typeName, string relationName, 
        Userset relationDefinition, List<RelationReference> typeRestrictions, DirectAssignmentValidator validator)
    {
        children = PrioritizeDirectAssignment(children ?? new List<Userset>());

        var formattedUnion = new StringBuilder();
        var notFirst = false;
        foreach (var child in children)
        {
            if (notFirst)
            {
                formattedUnion.Append($" {op} ");
            }
            else
            {
                notFirst = true;
            }
            formattedUnion.Append(FormatSubRelation(typeName, relationName, child, typeRestrictions, validator));
        }

        return formattedUnion;
    }

    private static List<Userset> PrioritizeDirectAssignment(List<Userset> usersets)
    {
        if (usersets.Count > 0)
        {
            var thisPosition = usersets.FindIndex(u => u.This != null);
            if (thisPosition > 0)
            {
                var thisUserset = usersets[thisPosition];
                usersets.RemoveAt(thisPosition);
                usersets.Insert(0, thisUserset);
            }
        }

        return usersets;
    }

    private class DirectAssignmentValidator
    {
        private int _occurred = 0;

        public void Incr()
        {
            _occurred++;
        }

        public int Occurrences()
        {
            return _occurred;
        }

        public bool IsFirstPosition(Userset userset)
        {
            if (userset.This != null)
            {
                return true;
            }

            if (userset.Difference?.Base != null)
            {
                if (userset.Difference.Base.This != null)
                {
                    return true;
                }
                else
                {
                    return IsFirstPosition(userset.Difference.Base);
                }
            }
            else if (userset.Intersection?.Child != null && userset.Intersection.Child.Count > 0)
            {
                if (userset.Intersection.Child[0].This != null)
                {
                    return true;
                }
                else
                {
                    return IsFirstPosition(userset.Intersection.Child[0]);
                }
            }
            else if (userset.Union?.Child != null && userset.Union.Child.Count > 0)
            {
                if (userset.Union.Child[0].This != null)
                {
                    return true;
                }
                else
                {
                    return IsFirstPosition(userset.Union.Child[0]);
                }
            }
            return false;
        }
    }

    private StringBuilder FormatSubRelation(string typeName, string relationName, Userset relationDefinition, 
        List<RelationReference> typeRestrictions, DirectAssignmentValidator validator)
    {
        if (relationDefinition.This != null)
        {
            validator.Incr();
            return new StringBuilder(FormatThis(typeRestrictions));
        }

        if (relationDefinition.ComputedUserset != null)
        {
            return new StringBuilder(FormatComputedUserset(relationDefinition));
        }

        if (relationDefinition.TupleToUserset != null)
        {
            return new StringBuilder(FormatTupleToUserset(relationDefinition));
        }

        if (relationDefinition.Union != null)
        {
            return FormatUnion(typeName, relationName, relationDefinition, typeRestrictions, validator)
                .Insert(0, '(')
                .Append(')');
        }

        if (relationDefinition.Intersection != null)
        {
            return FormatIntersection(typeName, relationName, relationDefinition, typeRestrictions, validator)
                .Insert(0, '(')
                .Append(')');
        }

        if (relationDefinition.Difference != null)
        {
            return FormatDifference(typeName, relationName, relationDefinition, typeRestrictions, validator)
                .Insert(0, '(')
                .Append(')');
        }

        throw new UnsupportedDSLNestingException(typeName, relationName);
    }

    private string FormatThis(List<RelationReference> typeRestrictions)
    {
        var restrictions = typeRestrictions ?? new List<RelationReference>();
        return "[" + string.Join(", ", restrictions.Select(FormatTypeRestriction)) + "]";
    }

    private string FormatTypeRestriction(RelationReference restriction)
    {
        var typeName = restriction.Type;
        var relation = restriction.Relation;
        var wildcard = restriction.Wildcard;
        var condition = restriction.Condition;

        var formattedTypeRestriction = new StringBuilder(typeName);

        if (wildcard != null)
        {
            formattedTypeRestriction.Append(":*");
        }

        if (!string.IsNullOrEmpty(relation))
        {
            formattedTypeRestriction.Append('#').Append(relation);
        }

        if (!string.IsNullOrEmpty(condition))
        {
            formattedTypeRestriction.Append(" with ").Append(condition);
        }
        
        return formattedTypeRestriction.ToString();
    }

    private string FormatComputedUserset(Userset relationDefinition)
    {
        return relationDefinition.ComputedUserset?.Relation ?? string.Empty;
    }

    private string FormatTupleToUserset(Userset relationDefinition)
    {
        var computedUserset = string.Empty;
        var tupleset = string.Empty;
        
        if (relationDefinition?.TupleToUserset != null)
        {
            if (relationDefinition.TupleToUserset.ComputedUserset != null)
            {
                computedUserset = relationDefinition.TupleToUserset.ComputedUserset.Relation ?? string.Empty;
            }
            if (relationDefinition.TupleToUserset.Tupleset != null)
            {
                tupleset = relationDefinition.TupleToUserset.Tupleset.Relation ?? string.Empty;
            }
        }
        
        return new StringBuilder(computedUserset).Append(" from ").Append(tupleset).ToString();
    }

    private string FormatConditions(AuthorizationModel model)
    {
        var conditions = model?.Conditions;
        if (conditions == null || conditions.Count == 0)
        {
            return string.Empty;
        }

        var formattedConditions = new StringBuilder();
        var sortedConditions = new SortedDictionary<string, Condition>(conditions);

        foreach (var conditionEntry in sortedConditions)
        {
            var conditionName = conditionEntry.Key;
            var conditionDef = conditionEntry.Value;

            var formattedCondition = FormatCondition(conditionName, conditionDef);
            formattedConditions.Append(EOL).Append(formattedCondition);
        }

        return formattedConditions.ToString();
    }

    private string FormatCondition(string conditionName, Condition conditionDef)
    {
        if (!conditionName.Equals(conditionDef.Name))
        {
            throw new ArgumentException("conditionName must match condition.Name()");
        }

        var formattedParameters = FormatConditionParameters(conditionDef.Parameters);
        return new StringBuilder("condition ")
            .Append(conditionDef.Name)
            .Append('(')
            .Append(formattedParameters)
            .Append(") {")
            .Append(EOL)
            .Append("  ")
            .Append(conditionDef.Expression)
            .Append(EOL)
            .Append('}')
            .Append(EOL)
            .ToString();
    }

    private string FormatConditionParameters(Dictionary<string, ConditionParamTypeRef>? parameters)
    {
        if (parameters == null || parameters.Count == 0)
        {
            return string.Empty;
        }

        var sortedParameters = new SortedDictionary<string, ConditionParamTypeRef>(parameters);
        return string.Join(", ", sortedParameters.Select(entry =>
        {
            var parameterName = entry.Key;
            var parameterType = entry.Value;
            var formattedParameterType = parameterType.TypeName.ToString()
                .Replace("TYPE_NAME_", "")
                .ToLowerInvariant();
            
            if (formattedParameterType == "list" || formattedParameterType == "map")
            {
                var genericTypeString = string.Empty;
                if (parameterType.GenericTypes != null && parameterType.GenericTypes.Count > 0)
                {
                    genericTypeString = parameterType.GenericTypes[0].TypeName.ToString()
                        .Replace("TYPE_NAME_", "")
                        .ToLowerInvariant();
                }
                formattedParameterType = $"{formattedParameterType}<{genericTypeString}>";
            }
            
            return $"{parameterName}: {formattedParameterType}";
        }));
    }
}
